### PR TITLE
#164956720 create user role and permission  management UI

### PR DESCRIPTION
**Pull request by AmosWels**
#### What does this PR do?
- Creates a user role and permission management UI
#### Description of Task to be completed?
- Adds a UI design for an Admin to create/update a role and assign different permissions to different roles from the UI designed.
- Adds a UI to enable an admin user to mark a role as Active, Inactive or Delete the role.
- Enables an admin user to view a list of all available roles.
#### How should this be manually tested?
- Fetch branch `ft-user-role-and-permisions-ui-164956720`
- Log into the application and access the Admin panel
- Click on the side nav bar click `Users`.
- Navigate to `Users Roles`.
- Perform CRUD operations on the users roles

#### What are the relevant pivotal tracker stories?
- [#164956720](https://www.pivotaltracker.com/story/show/164956720)

#### Background Context
- An admin wasn't able to create/update a role and assign different permissions to different roles from the UI

#### Screenshots (If Applicable)
image

#### Gif (If Applicable)
user roles

### CODE REVIEW

**Comment by AmosWels** [comment]
> #### What does this PR do?
> * Creates a user role and permission management UI
> 
> #### Description of Task to be completed?
> * Adds a UI design for an Admin to create/update a role and assign different permissions to different roles from the UI designed.
> * Adds a UI to enable an admin user to mark a role as Active, Inactive or Delete the role.
> * Enables an admin user to view a list of all available roles.
> 
> #### How should this be manually tested?
> * Fetch branch `ft-user-role-and-permisions-ui-164956720`
> * Log into the application and access the Admin panel
> * Click on the side nav bar click `Users`.
> * Navigate to `Users Roles`.
> * Perform CRUD operations on the users roles
> 
> #### What are the relevant pivotal tracker stories?
> * [#164956720](https://www.pivotaltracker.com/story/show/164956720)
> 
> #### Background Context
> N/A
> 
> #### Screenshots (If Applicable)
> 
image




> #### What does this PR do?
> * Creates a user role and permission management UI
> 
> #### Description of Task to be completed?
> * Adds a UI deaign for an Admin to create/update a role and assign different permissions to different roles from the UI designed.
> * Adds a UI to enable an admin user to mark a role as Active, Inactive or Delete the role.
> * Enables an admin user to view a list of all available roles.
> 
> #### How should this be manually tested?
> * Fetch branch `ft-user-role-and-permisions-ui-164956720`
> * Log into the application and access the Admin panel
> * Click on the side nav bar click `Users`.
> * Navigate to `Users Roles`.
> * Perform CRUD operations on the users roles
> 
> #### What are the relevant pivotal tracker stories?
> * [#164956720](https://www.pivotaltracker.com/story/show/164956720)
> 
> #### Background Context
> N/A
> 
> #### Screenshots (If Applicable)
> image

**Comment by sulenchy** [comment]
@AmosWels You are having conflicting files on this PR. Could you please resolve the conflict anytime soon?

File: client/src/components/Admin/Tabs/Users.jsx [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        <Users />
+      </div>
+      <div label="User Roles">
+        <UserRoles />
+      </div>
+    </Tabs>
+  </div>
+);
+
+export default Index;
```
Comment (Yiga-fred) [inline client/src/components/Admin/Tabs/Users.jsx]: Missing a new line here

File: client/src/components/Admin/Users/UserRoleCard.jsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  role: PropTypes.shape({
+    endDate: PropTypes.string,
+    startDate: PropTypes.string,
+  }),
+  showDeleteModal: PropTypes.func,
+  showPermisionModal: PropTypes.func,
+  showEditModal: PropTypes.func,
+};
+
+export default RolesCard;
```
Comment (Yiga-fred) [inline client/src/components/Admin/Users/UserRoleCard.jsx]: New line here too

File: client/src/tests/components/Admin/Tabs/Users.test.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+import { shallow } from 'enzyme';
+import Index from '../../../../components/Admin/Tabs/Users';
+
+describe('Users Tab Component', () => {
+  it('should render correctly', () => {
+    const wrapper = shallow(<Index />);
+    expect(wrapper).toMatchSnapshot();
+    expect(wrapper.length).toEqual(1);
+  });
+});
```
Comment (Yiga-fred) [inline client/src/tests/components/Admin/Tabs/Users.test.js]: New line

File: client/src/components/App.jsx [hunk 4]
```diff
@@ -13,13 +13,14 @@ import ExportOrders from './Admin/OrderHistory/ExportOrders';
 import Meals from './Admin/Meals/Index';
 import Engagements from './Admin/Engagements/Index';
 import Ratings from './Admin/Ratings/Index';
-import Users from './Admin/Users/Users';
+// import Users from './Admin/Users/Users';
```
Comment (zellymiriam) [inline client/src/components/App.jsx]: remove commented code

File: client/src/components/Admin/Users/UserRoles.jsx [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        />
+        <DeleteModal
+          deleteItem={this.deleteRole}
+          displayDeleteModal={displayDeleteModal}
+          closeModal={this.closeModal}
+          modalContent={modalContent}
+          item="Role"
+        /> 
+        <PermissionModal
+          // deleteItem={this.deleteRole}
```
Comment (zellymiriam) [inline client/src/components/Admin/Users/UserRoles.jsx]: remove commented code

File: client/src/components/Admin/Users/UserRoles.jsx [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+          {roles.length > 0 && this.renderRoles(roles)}
+        </div>
+        <ToastContainer />
+        <Modal
+          displayModal={displayModal}
+          closeModal={this.closeModal}
+          handleSubmit={this.handleSubmit}
+          onChange={this.onChange}
+          name={name}
+          // formValidation={this.formValidation}
```
Comment (zellymiriam) [inline client/src/components/Admin/Users/UserRoles.jsx]: remove comment

File: client/src/components/Admin/Users/UserRoleCard.jsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import moment from 'moment';
+import dateFormatter from '../../../helpers/dateFormatter';
+
+
```
Comment (zellymiriam) [inline client/src/components/Admin/Users/UserRoleCard.jsx]: Please remove this line.

File: client/src/components/Admin/Tabs/Users.jsx (hunk 1 above)
Comment (AmosWels) [reply client/src/components/Admin/Tabs/Users.jsx]: New line has been added

File: client/src/components/Admin/Users/UserRoleCard.jsx (hunk 2 above)
Comment (AmosWels) [reply client/src/components/Admin/Users/UserRoleCard.jsx]: New line has been added

File: client/src/tests/components/Admin/Tabs/Users.test.js (hunk 3 above)
Comment (AmosWels) [reply client/src/tests/components/Admin/Tabs/Users.test.js]: New line has been added

File: client/src/components/App.jsx (hunk 4 above)
Comment (AmosWels) [reply client/src/components/App.jsx]: commented code has been removed

File: client/src/components/Admin/Users/UserRoleCard.jsx (hunk 7 above)
Comment (AmosWels) [reply client/src/components/Admin/Users/UserRoleCard.jsx]: extra blank line has been removed

**Comment by SoniaRMK** [comment]
@AmosWels kudos on the good work. I have one comment about how permissions are assigned to roles. Currently, if I need to add say 5 permissions to a role, I would have to do that five times aka only one permission can be added at a time. However, in order to improve user experience, this should all be done within one click. Kindly refer to how side meals are added to a menu for clarity.

**Comment by AmosWels** [comment]
> @AmosWels kudos on the good work. I have one comment about how permissions are assigned to roles. Currently, if I need to add say 5 permissions to a role, I would have to do that five times aka only one permission can be added at a time. However, in order to improve user experience, this should all be done within one click. Kindly refer to how side meals are added to a menu for clarity.

Thanks, @SoniaRMK for the feedback. The constraint of only one permission to a role at a time comes from how the backend was designed. This approach was after the syncs I had with the backend team.

**Pull request merged**